Initial implementation of dynamic test generation

paul-tqh-nguyen opened the pull request:
This PR contains an initial implementation of dynamic test generation for our `graphblas-opt` tests.

It provides infrastructure for writing parameterized tests that generate `*mlir` test files at test-running time.

This is intended to address https://github.com/metagraph-dev/mlir-graphblas/issues/66.

paul-tqh-nguyen commented:
I've made several changes in the most recent commits to have the dynamically generated tests be run via `pytest` instead of at `graphblas-opt` compile time. 

This commit also makes it so that `pytest` by default runs with `-n auto` so that the tests can be run in parallel. 

I think this PR is ready for a second review.

jim22k commented on conftest.py:
```diff
@@ -1,14 +1,330 @@
+                    f"{repr(parameter_value_choices)} does not specify a valid set of parameter values."
+                )
+
+        # Handle each sampling case separately
```
Which sampling method matches the existing behavior of one value per parameter with a predictable choice?
`sampling=1` seems the closest, but the choice of the value is random.

paul-tqh-nguyen replied:
I don't think I have one for what you're describing. I can add one if desired. 

I'll summarize what I have so far.

Bucket:
```
for each template
    for each parameter
        for each possible value this parameter can take:
            randomly pick exactly one value for all the other parameters
            Make a test
```

Exhaustive:
```
for each template
    grab all possible tests for this template
    add all the possible tests to the final tests
```

If `sampling` is an `int`:
```
for each template:
    grab all possible tests for this template
    randomly "sampling" number of tests from all possible tests (add these to the final tests)
```

If `sampling` is a `float` in the range [0,1]:
```
for each template:
    grab all possible tests for this template
    randomly "sampling" percent of all possible tests (add these to the final tests)
```

I think we can have one for what you're describing and call it `default` that'll work like this:
```
for each template:
    for each parameter
        pick the first choice (deterministic since the parameter choices are lists).
    add exactly one test for this template
```

jim22k replied:
I like `default` for the name. Can we also make that the default method of running the tests if not specified?

paul-tqh-nguyen replied:
I've implemented the `default` option and made it the default method of running tests. I've broken the tests somewhere in there, so I'll be fixing that.

paul-tqh-nguyen commented on conftest.py:
```diff
@@ -1,14 +1,330 @@
+        sampling = float(sampling_method_string)
+    else:
+        sampling = sampling_method_string
+    seed = (
```
I should probably add a code comment here saying what this is doing. 

All the workers need to have run the exact same tests. Thus, they need to use the exact same random seed when sampling the tests (all the workers redundantly do this; not sure how to avoid this redundant work). I do this by making the `testrunuid` the seed (via all the `zlib.adler32` stuff) if the worker is running this code. Since the main/non-worker process doesn't run any tests (so nothing bad happens if we randomly sample different tests) and since the main process's `metafunc.config` doesn't have a `workerinput` attribute, I just use `time.time()`.

jim22k replied:
Agree that it's a bit weird, but as long as it's well documented for *why* it behaves this way and what it attempting to accomplish, I'm fine with it. Probably just copy-paste exactly what you wrote here and put it in a comment in the code.

paul-tqh-nguyen replied:
I added the comment.

paul-tqh-nguyen committed:
Initial implementation of dynamic test generation

This commit contains the initial implementation of dynamic test generation for our graphblas-opt tests.

We now have a framework for writing parameterized tests that generate *mlir files at test-running time.
